File: cases/taint/scala/nested-access-path-positive/NestedAccessPathPositive.scala
package dataflowbench

object NestedAccessPathPositive {
  class Inner {
    var value: String = "clean"
    var other: String = "clean"
  }

  class Middle {
    val inner: Inner = new Inner()
  }

  class Outer {
    val middle: Middle = new Middle()
  }

  def dfb_source(): String = { // DFB-SOURCE: nested-access-path-input
    "tainted"
  }

  def dfb_sink(value: String): Unit = {} // DFB-SINK: nested-access-path-sink

  def run(): Unit = {
    val outer = new Outer()
    outer.middle.inner.value = dfb_source() // DFB-WITNESS: nested-access-path-store
    dfb_sink(outer.middle.inner.value)
  }
}

File: cases/taint/cpp/nested-access-path-positive/nested_access_path_positive.cpp
#include <string>

std::string dfb_source() { // DFB-SOURCE: nested-access-path-input
    return "tainted";
}

void dfb_sink(const std::string &value) {} // DFB-SINK: nested-access-path-sink

struct Leaf {
    std::string value;
    std::string other;
};

struct Middle {
    Leaf c;
};

struct Root {
    Middle b;
};

void run() {
    Root a;
    a.b.c.value = dfb_source(); // DFB-WITNESS: nested-access-path-store
    a.b.c.other = "clean";
    dfb_sink(a.b.c.value);
}
